File: AjarinAndroid/src/main/java/com/example/ajarin/android/landing/presentation/register/sections/CreateAccount.kt
package com.example.ajarin.android.landing.presentation.register.sections

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.example.ajarin.android.core_ui.theme.AjarinTheme
import com.example.ajarin.android.core_ui.components.BasicTextField
import com.example.ajarin.android.core_ui.components.PasswordTextField
import com.example.ajarin.android.core_ui.components.PrimaryButton
import com.example.ajarin.landing.presentation.register.RegisterEvent
import com.example.ajarin.landing.presentation.register.RegisterSection
import com.example.ajarin.landing.presentation.register.RegisterState

@Composable
fun CreateAccount(
    state: RegisterState,
    onEvent: (RegisterEvent) -> Unit,
    onSignIn: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Spacer(modifier = Modifier.height(64.dp))

        Column(
            modifier = Modifier
                .fillMaxWidth(),
            horizontalAlignment = Alignment.Start
        ) {
            Text(
                text = "Create Account",
                style = MaterialTheme.typography.h5
            )

            Spacer(modifier = Modifier.height(4.dp))

            Text(
                text = "Create an account to access all features in this app.",
                style = MaterialTheme.typography.body2.copy(
                    color = Color.Gray
                )
            )
        }

        Spacer(modifier = Modifier.height(16.dp))

        BasicTextField(
            label = "Username",
            text = state.name,
            onTextChange = {
                onEvent(
                    RegisterEvent.OnNameChange(it)
                )
            },
            error = state.nameError
        )

        Spacer(modifier = Modifier.height(8.dp))

        BasicTextField(
            label = "Email",
            text = state.email,
            onTextChange = {
                onEvent(
                    RegisterEvent.OnEmailChange(it)
                )
            },
            error = state.emailError
        )

        Spacer(modifier = Modifier.height(8.dp))

        PasswordTextField(
            password = state.password,
            onPasswordChange = {
                onEvent(
                    RegisterEvent.OnPasswordChange(it)
                )
            },
            passwordVisible = state.isPasswordVisible,
            showPassword = {
                onEvent(
                    RegisterEvent.ToggleShowPassword
                )
            },
            error = state.passwordError
        )

        Spacer(modifier = Modifier.height(16.dp))

        PrimaryButton(
            text = "Next",
            textModifier = Modifier
                .fillMaxWidth(),
            isEnabled = state.registerEnabled
        ) {
            onEvent(
                RegisterEvent.UpdateSection(RegisterSection.FillNumber)
            )
        }

        Spacer(modifier = Modifier.height(16.dp))

        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.Center
        ) {
            Text(
                text = "Already have an account?",
                style = MaterialTheme.typography.body2.copy(
                    color = Color.Gray
                )
            )

            Spacer(modifier = Modifier.width(2.dp))

            Text(
                text = "Sign In",
                style = MaterialTheme.typography.body2.copy(
                    color = MaterialTheme.colors.primary
                ),
                modifier = Modifier
                    .clickable {
                        onSignIn()
                    }
            )
        }
    }
}

@Preview
@Composable
fun CreateAccountPreview() {
    AjarinTheme {
        CreateAccount(
            state = RegisterState(),
            onEvent = {  },
            onSignIn = {  }
        )
    }
}

File: AjarinIos/AjarinIos/landing/presentation/register/sections/CreateAccount.swift
import SwiftUI
import shared

struct CreateAccount: View {
    let state: RegisterState
    let onEvent: (RegisterEvent) -> Void
    @Environment(\.presentationMode) var presentationMode
    
    var body: some View {
        VStack {
            Spacer()
            
            HStack {
                VStack(alignment: .leading) {
                    Text("Create Account")
                        .font(.title2)
                    
                    Spacer()
                        .frame(height: 2)
                    
                    Text("Create an account to access all features in this app.")
                        .font(.subheadline)
                }
                
                Spacer()
            }
            
            Spacer()
                .frame(height: 32)
            
            Group {
                DefaultTextField(
                    title: "Username",
                    text: Binding(
                        get: {
                            state.name
                        },
                        set: {
                            onEvent(
                                .OnNameChange(name: $0)
                            )
                        }
                    )
                )
                
                if let nameError = state.nameError {
                    HStack {
                        Text(nameError.description())
                            .font(.caption)
                            .foregroundColor(.red)
                        
                        Spacer()
                    }
                }
                
                Spacer()
                    .frame(height: 8)
                
                DefaultTextField(
                    title: "Email",
                    text: Binding(
                        get: {
                            state.email
                        },
                        set: {
                            onEvent(
                                .OnEmailChange(email: $0)
                            )
                        }
                    )
                )
                
                if let emailError = state.emailError {
                    HStack {
                        Text(emailError.description())
                            .font(.caption)
                            .foregroundColor(.red)
                        
                        Spacer()
                    }
                }
                
                Spacer()
                    .frame(height: 8)
                
                PasswordTextField(
                    hint: "Password",
                    passwordVisible: Binding(
                        get: {
                            state.isPasswordVisible
                        },
                        set: {_ in
                            onEvent(
                                .ToggleShowPassword()
                            )
                        }
                    ),
                    text: Binding(
                        get: {
                            state.password
                        },
                        set: {
                            onEvent(
                                .OnPasswordChange(password: $0)
                            )
                        }
                    )
                )
                
                if let passwordError = state.passwordError {
                    HStack {
                        Text(passwordError.description())
                            .font(.caption)
                            .foregroundColor(.red)
                        
                        Spacer()
                    }
                }
            }
            
            Spacer()
                .frame(height: 32)
            
            Button {
                onEvent(
                    .UpdateSection(section: .fillnumber)
                )
            } label: {
                Text("Next")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            .disabled(!state.registerEnabled)
            
            HStack {
                Text("Already have an account?")
                
                Button {
                    presentationMode.wrappedValue.dismiss()
                } label: {
                    Text("Sign In")
                }
            }
            
            Spacer()
        }
        .padding()
    }
}

struct CreateAccount_Previews: PreviewProvider {
    static var previews: some View {
        CreateAccount(
            state: RegisterState(
                currentSection: .createaccount,
                name: "",
                nameError: nil,
                email: "",
                emailError: nil,
                password: "",
                passwordError: nil,
                isPasswordVisible: false,
                registerEnabled: false,
                number: "",
                numberError: nil,
                sendOtpEnabled: false,
                otp: "",
                otpError: nil,
                verifyOtpEnabled: false,
                otpSendCount: 0,
                otpCountDown: 0,
                verificationId: nil,
                sendOtpError: nil,
                sendOtpLoading: false,
                verifyOtpError: nil,
                verifyOtpLoading: false,
                registerLoading: false,
                registerError: nil,
                registerSuccess: false
            ),
            onEvent: { _ in }
        )
    }
}
